File: template/app/src/main/java/com/xmartlabs/template/ui/onboarding/OnboardingPresenter.java
package com.xmartlabs.template.ui.onboarding;

import com.annimon.stream.Optional;
import com.xmartlabs.bigbang.ui.mvp.BaseMvpPresenter;

import lombok.Builder;

@Builder
public class OnboardingPresenter extends BaseMvpPresenter<OnboardingView> {
  private static final int FIRST_PAGE = 0;

  private OnboardingPageAdapter pageAdapter;
  @Builder.Default
  private int currentPage = FIRST_PAGE;

  @Override
  public void attachView(OnboardingView view) {
    super.attachView(view);
    pageAdapter = view.createPageAdapter();
    view.setupView(pageAdapter);
  }

  void nextButtonPressed() {
    if (isLastPage()) {
      skipButtonPressed();
    } else {
      currentPage++;
      Optional.ofNullable(getView())
          .ifPresent(view -> view.moveToPage(currentPage));
      handleSkipButtonVisibility();
    }
  }

  public boolean isLastPage() {
    return currentPage == pageAdapter.getCount() - 1;
  }

  private void handleSkipButtonVisibility() {
    Optional.ofNullable(getView())
        .ifPresent(view -> view.setSkipButtonVisibility(!isLastPage()));
  }

  void pageChanged(int newPage) {
    currentPage = newPage;
    Optional.ofNullable(getView())
        .ifPresent(OnboardingView::handleNextButtonVisibility);
    handleSkipButtonVisibility();
  }

  void skipButtonPressed() {
    Optional.ofNullable(getView())
        .ifPresent(OnboardingView::goToLoginActivity);
  }
}

File: template/app/src/main/java/com/xmartlabs/template/ui/onboarding/OnboardingPresenter.kt
package com.xmartlabs.template.ui.onboarding

import com.xmartlabs.template.ui.common.TemplatePresenter
import javax.inject.Inject

class OnboardingPresenter @Inject constructor() : TemplatePresenter<OnboardingView>() {
  companion object {
    private val FIRST_PAGE = 0
  }

  val isLastPage: Boolean
    get() = currentPage == pageAdapter.count - 1

  @Suppress("LateinitUsage")
  private lateinit var pageAdapter: OnboardingPageAdapter
  private var currentPage = FIRST_PAGE

  override fun attachView(view: OnboardingView) {
    super.attachView(view)
    pageAdapter = view.createPageAdapter()
    view.setup(pageAdapter)
  }

  internal fun nextButtonPressed() {
    if (isLastPage) {
      skipButtonPressed()
    } else {
      currentPage++
      view?.moveToPage(currentPage)
      handleSkipButtonVisibility()
    }
  }

  private fun handleSkipButtonVisibility() = view?.setSkipButtonVisibility(!isLastPage)

  internal fun pageChanged(newPage: Int) {
    currentPage = newPage
    view?.handleNextButtonVisibility()
    handleSkipButtonVisibility()
  }

  internal fun skipButtonPressed() = view?.goToLoginActivity()
}
